### ISSUE TITLE
main_supcon_efficientnet，loss不下降

### ISSUE THREAD

**Issue by depblack** [xuyingzhongguo/deepfake_supcon#2]
1. 作者您推荐的数据预处理代码（有一些视频并没有提取30帧，这其中绝大多数是28~29帧，也有极少数是4帧/1帧的情况）；
2. 使用的命令是，python main_supcon_efficientnet.py --print_freq 1 --batch_size 32 --learning_rate 0.001 --dataset deepfakes --temp 0.1 --cosine ； 其中代码默认使用的MyDataset_wavelet，困惑但没有改动。
3. 具体训练输出如下：
Train: [2][1516/3324]   BT 0.259 (5.902)        DT 0.001 (5.693)        loss 5.919 (5.919)
torch.Size([64, 1280, 1, 1])
torch.Size([64, 1280])
torch.Size([64, 128])
loss一直保持不下降，而我除了数据集（train.txt）之外没有改变（default = ‘*/train.txt’）

希望能得到解答，难道说其实使用的是main_supcon.py？

**Comment by xuyingzhongguo** [comment]
1. The reason might be the face recognition confidence [threshold ](https://github.com/xuyingzhongguo/Deepfake-face-extraction/blob/main/FF%2B%2B/extract_faces.py#:~:text=if%20box%20is%20not%20None%20and%20float(prob%5B0%5D)%20%3E%200.99%3A) is so high that it does not extract those faces for training purpose. I set it to 0.99 but you can try with lower values. Otherwise I would suspect the video is less than 300 frames.
2. I think MyDataset would be the correct one. I will update the code. Thank you for telling. (The wavelet one is just a try.)
3. ~~I think efficientnet would be the one I used.~~ I think I used xception. It is in the networks folder. You could invoke it when loading the model. (Sorry I forgot about everything.) How many epochs did you train? Have you tried different learning rate?

(Sorry I don't have Chinese input on this laptop =.=)

**Comment by depblack** [comment]
感谢回复！
1.1 是的，阈值设置的比较高，似乎是0.99+的程度，没有去确认。此外它的设置只是“提取每个视频前300帧的30帧”，大概就是1，11，21，...。这样每个视频提取大概30帧。&nbsp;
1.2 我直接使用了源代码没有修改，因为您在issue 1种回复并介绍您这篇工作的实验设置是“每个视频提取了10帧（训练、测试）”(1.2 I extracted a frame every 10 frames for both training and test.), 我推测val可能多一些，不过也许30帧~100帧也够了？&nbsp;&nbsp;
1.3 但上述描述和您现在回复的“&nbsp;is less than 300 frames.” 似乎有一定的矛盾之处。
2. 好的
3. 我做了两点尝试
3.1 仅修改“每个视频选择多少帧”，（50，30，10），训练大概2~5个epoch后，回来观察发现基本在第一个epoch之后loss就不再下降
3.2 目前增大学习率尝试中 （0.001 -&gt; 0.01），也增大batch_size尝试中（32-&gt;64；而实际上TwoTransform之后会是64-&gt;128），这个操作来源是我去看了SupCon的github issue，原版ImageNet使用的应该是（batch_size:1024, learning_rate:0.05），而deepfake_FF++使用（32，0.001）似乎太小了，即使他们说loss不下降的解决方法是（减少batch_size以增大variation，减少学习率以更容易走出鞍点），我还是尝试了一下这样的设置（64, 0.01）


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "xuyingzhongguo/deepfake_supcon"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年4月3日(星期一) 下午3:11
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [xuyingzhongguo/deepfake_supcon] main_supcon_efficientnet，loss不下降 (Issue #2)





  
The reason might be the face recognition confidence threshold  is so high that it does not extract those faces for training purpose. I set it to 0.99 but you can try with lower values. Otherwise I would suspect the video is less than 300 frames.
 
I think MyDataset would be the correct one. I will update the code. Thank you for telling. (The wavelet one is just a try.)
 
I think efficientnet would be the one I used. (Sorry I forgot about everything.) How many epochs did you train? Have you tried different learning rate?
  
(Sorry I don't have Chinese input on this laptop =.=)

**Comment by depblack** [comment]
> 1. The reason might be the face recognition confidence [threshold ](https://github.com/xuyingzhongguo/Deepfake-face-extraction/blob/main/FF%2B%2B/extract_faces.py#:~:text=if%20box%20is%20not%20None%20and%20float(prob%5B0%5D)%20%3E%200.99%3A) is so high that it does not extract those faces for training purpose. I set it to 0.99 but you can try with lower values. Otherwise I would suspect the video is less than 300 frames.
> 2. I think MyDataset would be the correct one. I will update the code. Thank you for telling. (The wavelet one is just a try.)
> 3. ~I think efficientnet would be the one I used.~ I think I used xception. It is in the networks folder. You could invoke it when loading the model. (Sorry I forgot about everything.) How many epochs did you train? Have you tried different learning rate?
> 
> (Sorry I don't have Chinese input on this laptop =.=)



> 1. The reason might be the face recognition confidence [threshold ](https://github.com/xuyingzhongguo/Deepfake-face-extraction/blob/main/FF%2B%2B/extract_faces.py#:~:text=if%20box%20is%20not%20None%20and%20float(prob%5B0%5D)%20%3E%200.99%3A) is so high that it does not extract those faces for training purpose. I set it to 0.99 but you can try with lower values. Otherwise I would suspect the video is less than 300 frames.
> 2. I think MyDataset would be the correct one. I will update the code. Thank you for telling. (The wavelet one is just a try.)
> 3. ~I think efficientnet would be the one I used.~ I think I used xception. It is in the networks folder. You could invoke it when loading the model. (Sorry I forgot about everything.) How many epochs did you train? Have you tried different learning rate?
> 
> (Sorry I don't have Chinese input on this laptop =.=)

感谢解惑，关于xception又有了新问题，
1. 没有对xception进行改造吗，就像 from networks.effnet_big import SupConEffNet， 其中的SupConEffNet一样？而是直接使用xception？ 
2. 使用的是，class Xception，而非class Xception_concat（），是吗？
3. def xception()使用此函数时加载有一些问题，就换用了其他人的代码如下（而我的问题是，应该都可以吧）： 
model_urls = {
    'xception':'https://www.dropbox.com/s/1hplpzet9d7dv29/xception-c0a72b38.pth.tar?dl=1'
}
def xception(num_classes=2, pretrained='imagenet'):
    model = Xception()
    if pretrained:
        settings = model_urls['xception']
        #settings = pretrained_settings['xception'][pretrained]
        #assert num_classes == settings['num_classes'], \
        #    "num_classes should be {}, but is {}".format(settings['num_classes'], num_classes)

        #model = Xception(num_classes=num_classes)
        #model.load_state_dict(model_zoo.load_url(settings['url']))
        model.load_state_dict(model_zoo.load_url(settings))

        #model.input_space = settings['input_space']
        #model.input_size = settings['input_size']
        #model.input_range = settings['input_range']
        #model.mean = settings['mean']
        #model.std = settings['std']

    # TODO: ugly
    model.last_linear = nn.Linear(in_features=2048, out_features=num_classes)
    del model.fc
    return model

**Comment by xuyingzhongguo** [comment]
- 没有对xception进行改造吗，就像 from networks.effnet_big import SupConEffNet， 其中的SupConEffNet一样？而是直接使用xception？
No because the output of xception is 2 classes. The purpose of modification of effnet_big is just to make the last layer fc to 2 classes.

- 使用的是，class Xception，而非class Xception_concat（），是吗？
Yes, the simplest one. And it is fine to use your own xception or other backbones.
- 1.2 我直接使用了源代码没有修改，因为您在issue 1种回复并介绍您这篇工作的实验设置是“每个视频提取了10帧（训练、测试）”(1.2 I extracted a frame every 10 frames for both training and test.), 我推测val可能多一些，不过也许30帧~100帧也够了？
Maybe it was bad expression. I mean "a frame every 10 frames" == "10-frame interval". But I think the amount of the training set should be enough. 
- 我做了两点尝试
3.1 仅修改“每个视频选择多少帧”，（50，30，10），训练大概2~5个epoch后，回来观察发现基本在第一个epoch之后loss就不再下降
3.2 目前增大学习率尝试中 （0.001 -&gt; 0.01），也增大batch_size尝试中（32-&gt;64；而实际上TwoTransform之后会是64-&gt;128），这个操作来源是我去看了SupCon的github issue，原版ImageNet使用的应该是（batch_size:1024, learning_rate:0.05），而deepfake_FF++使用（32，0.001）似乎太小了，即使他们说loss不下降的解决方法是（减少batch_size以增大variation，减少学习率以更容易走出鞍点），我还是尝试了一下这样的设置（64, 0.01）
 just let me know if you have more questions.

**Comment by depblack** [comment]
> * 没有对xception进行改造吗，就像 from networks.effnet_big import SupConEffNet， 其中的SupConEffNet一样？而是直接使用xception？
>   No because the output of xception is 2 classes. The purpose of modification of effnet_big is just to make the last layer fc to 2 classes.
> * 使用的是，class Xception，而非class Xception_concat（），是吗？
>   Yes, the simplest one. And it is fine to use your own xception or other backbones.
> * 1.2 我直接使用了源代码没有修改，因为您在issue 1种回复并介绍您这篇工作的实验设置是“每个视频提取了10帧（训练、测试）”(1.2 I extracted a frame every 10 frames for both training and test.), 我推测val可能多一些，不过也许30帧~100帧也够了？
>   Maybe it was bad expression. I mean "a frame every 10 frames" == "10-frame interval". But I think the amount of the training set should be enough.
> * 我做了两点尝试
>   3.1 仅修改“每个视频选择多少帧”，（50，30，10），训练大概2~5个epoch后，回来观察发现基本在第一个epoch之后loss就不再下降
>   3.2 目前增大学习率尝试中 （0.001 -> 0.01），也增大batch_size尝试中（32->64；而实际上TwoTransform之后会是64->128），这个操作来源是我去看了SupCon的github issue，原版ImageNet使用的应该是（batch_size:1024, learning_rate:0.05），而deepfake_FF++使用（32，0.001）似乎太小了，即使他们说loss不下降的解决方法是（减少batch_size以增大variation，减少学习率以更容易走出鞍点），我还是尝试了一下这样的设置（64, 0.01）
>   just let me know if you have more questions.

非常感谢您的详细解答，目前没有其他困惑了，也许应当暂时Close this issue，下次再请教您
